[HLSL][DirectX] Extract HLSLBinding out of DXILResource. NFC

bogner opened the pull request:
We extract the binding logic out of the DXILResource analysis passes into the
FrontendHLSL library. This will allow us to use this logic for resource and
root signature bindings in both the DirectX backend and the HLSL frontend.

bogner committed:
[HLSL][DirectX] Extract HLSLBinding out of DXILResource. NFC

We extract the binding logic out of the DXILResource analysis passes into the
FrontendHLSL library. This will allow us to use this logic for resource and
root signature bindings in both the DirectX backend and the HLSL frontend.

inbelic commented on llvm/lib/Frontend/HLSL/HLSLBinding.cpp:
```diff
@@ -0,0 +1,141 @@
+  // sort all the collected bindings
+  llvm::stable_sort(Bindings);
+
+  // remove duplicates
```
Did we find a case that duplicates will be generated?

hekota replied:
The resource binding analysis in `DXILResourceBindingInfo::populate` calls tze builder's `trackBinding` for each `handlefrombinding` call it finds in the code and there can by multiple such calls for the same resource.

inbelic commented on llvm/lib/Frontend/HLSL/HLSLBinding.cpp:
```diff
@@ -0,0 +1,141 @@
+      S = &BS->Spaces.emplace_back(B.Space);
+
+    // The space is full - there are no free slots left, or the rest of the
+    // slots are taken by an unbounded array. Set flag to report overlapping
+    // binding later.
```
Is the last sentence an artifact from a previous implementation? IIUC, we will just invoke `ReportOverlap` rather than store a flag

bogner replied:
Yep, this is misleading now. I'll update the comment.

joaosaffran commented on llvm/lib/Frontend/HLSL/HLSLBinding.cpp:
```diff
@@ -0,0 +1,141 @@
+
+BindingInfo::RegisterSpace &
+BindingInfo::BindingSpaces::getOrInsertSpace(uint32_t Space) {
+  for (auto *I = Spaces.begin(); I != Spaces.end(); ++I) {
```
nit: https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop
```suggestion
  for (auto *I = Spaces.begin(), E = Spaces.end(); I != E; ++I) {
```

bogner replied:
Done. I also renamed the iterator to `It` and dropped the pointer type to make it clearer we're using an iterator interface here - there's no need to leak the information that `SmallVector<T>::iterator` is `T *` to the caller like this.

joaosaffran reviewed (approved):
Small nit, LGTM

hekota reviewed (commented):
LGTM!

bogner commented on llvm/include/llvm/Frontend/HLSL/HLSLBinding.h:
```diff
@@ -0,0 +1,162 @@
+  struct RegisterSpace {
+    uint32_t Space;
+    SmallVector<BindingRange> FreeRanges;
+    RegisterSpace(uint32_t Space) : Space(Space) {
+      FreeRanges.emplace_back(0, UINT32_MAX);
```
All of the `UINT32_MAX` uses are in code that was moved verbatim. That said, I do think `~0u` makes the intent clearer here, I'll update it.

bogner committed:
fixup: Clean up uses of UINT32_MAX and INT32_MAX

Use ~0u for the end-of-range sigil, and numeric_limits for math involving the
limits of uint32_t and int32_t.

bogner committed:
Improve overlap reporting related comments

bogner committed:
clean up iterator usage

bogner committed:
clang-format
